plugins/rest/azure_test.go

```diff
@@ -0,0 +1,157 @@
+package rest
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/open-policy-agent/opa/keys"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAzureManagedIdentitiesAuthPlugin_NewClient(t *testing.T) {
+	tests := []struct {
+		label      string
+		endpoint   string
+		apiVersion string
+		resource   string
+		objectID   string
+		clientID   string
+		miResID    string
+	}{
+		{
+			"test all defaults",
+			"", "", "", "", "", "",
+		},
+		{
+			"test no defaults",
+			"some_endpoint", "some_version", "some_resource", "some_oid", "some_cid", "some_miresid",
+		},
+	}
+
+	nonEmptyString := func(value string, defaultValue string) string {
+		if value == "" {
+			return defaultValue
+		}
+		return value
+	}
+
+	for _, tt := range tests {
+		config := generateConfigString(tt.endpoint, tt.apiVersion, tt.resource, tt.objectID, tt.clientID, tt.miResID)
+
+		client, err := New([]byte(config), map[string]*keys.Config{})
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		ap := client.config.Credentials.AzureManagedIdentity
+		_, err = ap.NewClient(client.config)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		// We test that default values are set correctly in the azureManagedIdentitiesAuthPlugin
+		// Note that there is significant overlap between TestAzureManagedIdentitiesAuthPlugin_NewClient and TestAzureManagedIdentitiesAuthPlugin
+		// This is because the latter cannot test default endpoint setting, which we do here
+		assert.Equal(t, nonEmptyString(tt.endpoint, azureIMDSEndpoint), ap.Endpoint, tt.label)
+		assert.Equal(t, nonEmptyString(tt.apiVersion, defaultAPIVersion), ap.APIVersion, tt.label)
+		assert.Equal(t, nonEmptyString(tt.resource, defaultResource), ap.Resource, tt.label)
+		assert.Equal(t, tt.objectID, ap.ObjectID, tt.label)
+		assert.Equal(t, tt.clientID, ap.ClientID, tt.label)
+		assert.Equal(t, tt.miResID, ap.MiResID, tt.label)
+	}
+}
+
+func TestAzureManagedIdentitiesAuthPlugin(t *testing.T) {
+	tests := []struct {
+		label      string
+		apiVersion string
+		resource   string
+		objectID   string
+		clientID   string
+		miResID    string
+		expected   string
+	}{
+		{
+			"test all defaults",
+			"", "", "", "", "",
+			"?api-version=2018-02-01&resource=https://storage.azure.com/",
```

Comment:
I find these tests rather difficult to comprehend. Could we rewrite them so that it's more obvious what's tested and what the expected outcome is?

Reply:
Good point, I've gone ahead and 
1. Renamed the field `expectedPath`
2. Represented the `expectedPath` in test cases by using `fmt.Sprintf` 
I hope this makes it a bit more readable. Feel free to resolve.

Reply:
Looks better 👍

Reply:
I've gone ahead and switched to testing with `url.Values` and `reflect.DeepEqual` as per https://github.com/open-policy-agent/opa/pull/3952#discussion_r741718120. 

That should fix this conversation in an even better way.